File: solutions/problem-12/solution.c
#include <stdio.h>
#include <math.h>

int divisors(long long int n){
	long long int i = 2, d = 1;

	while(i <= sqrt(n)){
		if(n % i == 0){
			++d;
			if(i != n / i) ++d;
		}
		++i;		
	}

	return d;
}

long long int triangle(long long int n){
	int sum = 0;

	for(long long int i = 1; i <= n; ++i) sum += i;

	return sum;
}

int main(){
	long long int i = 2, t = 1;
	int d = 0;

	while(1){
		d = divisors(t);
		if(d >= 501) break;
		t = triangle(i);
		++i;
	}

	printf("%lld\t%d\n", t, d);
	return 0;
}

File: solutions/problem-12/solution.py
from math import *

def divisors(n):
    limit = int(sqrt(n))
    divisors_list = []
    for i in range(1, limit+1, 1):
        if n % i == 0:
            divisors_list.append(i)
            if i != n/i:
                divisors_list.append(n/i)
    return len(divisors_list)
                
def isTriangleNumber(n):
    a = int(sqrt(2*n))
    return 0.5*a*(a+1) == n

def lastTerm(n):
    if isTriangleNumber(n):
        return int(sqrt(2*n))
    else:
        return None

check = 2**4 * 3**4 * 5**4 * 7 * 11

while not isTriangleNumber(check):
    check += 1
seriesLastTerm = lastTerm(check)

while divisors(check) <= 500:
    check += (seriesLastTerm + 1)
    seriesLastTerm += 1
print check
